be2-scala/src/main/scala/ch/epfl/pop/authentication/QRCodeChallengeGenerator.scala

```diff
@@ -0,0 +1,76 @@
+package ch.epfl.pop.authentication
+
+import akka.http.scaladsl.model.{ContentTypes, HttpEntity, ResponseEntity}
+import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
+import com.google.zxing.qrcode.encoder.{ByteMatrix, Encoder}
+
+import scala.io.Source
+
+/** Generates an http response that holds a web page with any desired challenge content in the form of a displayed qrcode The html/svg representation of the qrcode is inspired from https://github.com/svg/svgo
+  */
+object QRCodeChallengeGenerator {
+
+  private val templateFileName = "src/main/web/AuthenticationPageTemplate.html"
+
+  private val webTemplateQRCodeTemplate = "{{QRCODE_PATTERN}}"
+
+  private val qrcodeTotalSize = 650
+  private val qrcodeMargin = 25
+
+  /** Generates an html response representing a web page with a qrcode holding the content provided
+    * @param content
+    *   data to insert in the qrcode
+    * @return
+    *   a web page in the form of an http-html response
+    */
+  def generateChallengeContent(content: String): ResponseEntity = {
+    val encodedContent = Encoder.encode(content, ErrorCorrectionLevel.H)
+    val htmlQRCode = fromMatrixToHTML(encodedContent.getMatrix)
+
+    val templateFile = Source.fromFile(templateFileName)
+    val lines = for {
+      line <- templateFile.getLines()
+      substitutedLine = line.replace(webTemplateQRCodeTemplate, htmlQRCode)
+    } yield substitutedLine
```

Comment:
This should be my last comment before approval, I believe you forgot to close the stream

Reply:
fixed !